### PR TITLE
Add JavaScript solution

### PR DESCRIPTION
**Pull request by martingaston**
Here's a JavaScript solution to the gilded kata, very keen to hear feedback on how to make this better 😁 🎉

### CODE REVIEW

File: javascript/src/gilded_rose.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+class Item {
+  constructor(name, sellIn, quality) {
+    this.name = name;
+    this.sellIn = sellIn;
+    this.quality = quality;
+  }
+}
+
+class Foo extends Item {
```
Comment (felipesere) [inline javascript/src/gilded_rose.js]: Hmmm 🤔 
Do we need `Foo` and below `Bar` for our production code?
I get the sense these are for tests only?

I suggest moving these files over to the test directory.

File: javascript/src/gilded_rose.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  }
+}
+
+class Shop {
+  constructor(items = []) {
+    this.items = items;
+  }
+
+  updateQuality() {
+    this.items = this.items.map((item, index) => updateItem(item));
```
Comment (felipesere) [inline javascript/src/gilded_rose.js]: Unused `index` :)

File: javascript/src/items/AgedBrie.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+const UpdateItem = require("./Item");
+const { qualityIncrement } = require("../utils/calcControl");
+
+class UpdateAgedBrie extends UpdateItem {
+  constructor(item) {
```
Comment (felipesere) [inline javascript/src/items/AgedBrie.js]: It might be worth looking into, but I don't _think_ you need the `constructor` and the `super` call if you don't do anything with them....

File: javascript/src/utils/calcControl.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+const qualityIncrement = (quality, amount = 1) =>
```
Comment (felipesere) [inline javascript/src/utils/calcControl.js]: This comment is meant for the file name:
I don't think `calculationControl` is the most appealing name. 
How about this for a little magic:
What if you called the file `quality.js`, and then exported a an object called `quality` that had to functions: `increment(..)` and `decrement(..)`.
That way, you get to write code like
`if (quality.increment(brieQuality, 12))` 
 
😄

File: javascript/src/utils/updateItem.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const { qualityIncrement, qualityDecrement } = require("./calcControl");
+
+const {
+  UpdateItem,
+  UpdateAgedBrie,
+  UpdateSulfuras,
+  UpdateBackstagePass
+} = require("../items");
+
+function updateItem(item) {
```
Comment (felipesere) [inline javascript/src/utils/updateItem.js]: This is pretty snazzy!

How about pulling the `itemLookup` table out of the function itself to make the actual code stand out a bit more 😸 

I know `hasOwnProperty` is important if you pull out things from JS object that you don't own. In this particular case the "distance" between where the object is defined and where it is used is minimal: you can literally look up :) (pun intended).

I think something like this could work:

```
const itemUpdater = itemLookup[item.name] || defaultUpdater

return itemUpdater(item).tick()
```

File: javascript/src/utils/calcControl.js (hunk 4 above)
Comment (felipesere) [reply javascript/src/utils/calcControl.js]: Also, there is no need for the `utils` module. Quality control is at the forefront what we at Gilded Rose INC do.

You are highly encouraged to elevate the idea of `quality` into its own file 🏆 .

File: javascript/src/gilded_rose.js (hunk 2 above)
Comment (martingaston) [reply javascript/src/gilded_rose.js]: Dang, I totally should have spotted that. Also don't need the anonymous function, either 🤦‍♂️

File: javascript/src/gilded_rose.js (hunk 1 above)
Comment (martingaston) [reply javascript/src/gilded_rose.js]: 🤦‍♂️

File: javascript/.gitignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+*.swp
+node_modules/
```
Comment (fabientownsend) [inline javascript/.gitignore]: In general, I would keep these change in a different pull request. The reason I would do that is to leave in my PR only change/file that I want people to review and provide feedback :)

File: javascript/__tests__/gilded_rose.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    const gildedRose = new Shop([
+      new Item("foo", fooSellIn, fooQuality),
+      new Item("bar", barSellIn, barQuality)
+    ]);
+
+    const items = gildedRose.updateQuality();
+
+    expect(items[0].sellIn).toEqual(0);
+    expect(items[1].quality).toEqual(1);
+  });
```
Comment (fabientownsend) [inline javascript/__tests__/gilded_rose.js]: Something that I would do here is to really align your tests with the description. For example, I am not sure that using two different items is really necessary. Also, I would try to make it as obvious as possible for the reader to understand the state before and after an update.

Here is for example:
``` javascript
it("should decrease item sellIn and quality at the end of each day", () => {
  const fooSellIn = 1;
  const fooQuality = 1;
  const originalItem = new Item("foo", fooSellIn, fooQuality)

  const gildedRose = new Shop([originalItem]);

  const [updatedFooItem] = gildedRose.updateQuality();

  expect(updatedFooItem.sellIn).toEqual(originalItem.sellIn - 1);
  expect(updatedFooItem.quality).toEqual(originalItem.quality - 1);
});
```

Comment (fabientownsend) [reply javascript/__tests__/gilded_rose.js]: This made me realise that `updateQuality` is maybe misleading as it not only change the quality. What if instead of `updateQuality` was called something like `dailyUpdate`

File: javascript/__tests__/gilded_rose.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    const fooSellIn = 1;
+    const fooQuality = 1;
+    const sulfurasSellIn = 50;
+    const sulfurasQuality = 80;
+
+    const gildedRose = new Shop([
+      new Item("foo", fooSellIn, fooQuality),
+      new Item(sulfuras, sulfurasSellIn, sulfurasQuality)
+    ]);
+    const items = gildedRose.updateQuality();
```
Comment (fabientownsend) [inline javascript/__tests__/gilded_rose.js]: You can use JavaScript deconstruction feature:
``` javascript
const [firstItem, secondItem] = gildedRose.updateQuality();
```

File: javascript/src/gilded_rose.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+const updateItem = require("./utils/updateItem");
+
+class Item {
```
Comment (fabientownsend) [inline javascript/src/gilded_rose.js]: Do you think of any way you could delete `Item` and use the concrete classes: `UpdateAgedBrie`, `UpdateBackstagePasses`, etc?

File: javascript/.gitignore (hunk 6 above)
Comment (martingaston) [reply javascript/.gitignore]: This is a good point, well made 👍

File: javascript/__tests__/gilded_rose.js (hunk 7 above)
Comment (martingaston) [reply javascript/__tests__/gilded_rose.js]: This is great feedback thank you 😁
